arwen/contexts/runtime.go

```diff
@@ -206,6 +250,55 @@ func (context *runtimeContext) makeInstanceFromContractByteCode(contract []byte,
 	return nil
 }
 
+func (context *runtimeContext) useWarmInstanceIfExists(gasLimit uint64, codeHash []byte, newCode bool) bool {
+	if newCode || len(codeHash) == 0 {
+		return false
+	}
+
+	if context.IsContractOnTheStack(context.scAddress) {
+		return false
+	}
+
+	cachedObject, ok := context.warmInstanceCache.Get(codeHash)
+	if !ok {
+		return false
+	}
+
+	cachedMemoryObject, ok := context.localDataCache.Get(codeHash)
+	if !ok {
+		return false
+	}
+
+	warmInstance, ok := cachedObject.(wasmer.InstanceHandler)
+	if !ok {
+		return false
+	}
+
+	localMemory, ok := cachedMemoryObject.([]byte)
+	if !ok {
+		return false
+	}
+
+	warmInstanceMemory := warmInstance.GetMemory().Data()
+	if len(warmInstanceMemory) != len(localMemory) {
+		// TODO shrink the instance memory instead and return true
+		return false
+	}
+
+	copy(warmInstanceMemory, localMemory)
```

Comment:
warmInstanceMemory (the destination on which the copy dumps the data) is not used in the rest of this function.

Reply:
but it is actually copied inside the memory space of the wasm code :dagger:

Reply:
we need to restore memory always as it was from first compilation